### PR TITLE
feat: new pkg for chains define

### PR DESCRIPTION
**Pull request by ChefJerry**
### Summary
📦🔗🌐

This pull request introduces a new package, `@pancakeswap/chains`, that provides chain-related information and utilities for Pancakeswap. The package depends on several libraries for blockchain development and uses `tsup`, `tsc`, and `vitest` tools for bundling, transpiling, and testing. The package exports chain objects, chain IDs, and an enum for the supported networks.

> _There once was a package called `chains`_
> _That handled the blockchain domains_
> _It used `viem` and `tsup`_
> _And `vitest` to test up_
> _The code that it exported and maintained_

### Walkthrough
*  Add a new package `@pancakeswap/chains` that provides chain-related information for Pancakeswap ([link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-80d2965179c23b75c372ee4754951caeb4fb7bc1e8ee739081a62144bb871cdfR1-R50), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-1fa09fcb0bd6f74b404abf016fe1af0a7527858afe19d4c3e791def9053e29acR1-R20), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-cb6590dddd8646cdd5e3677cb1e05aa883274f7c7fd77fcd5645ac87bbf5713dR1-R20), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-fa0ddb5c9e072389f95a226dc82d6a9d75c7ce4bf6a8a2367a0fcf0ded7f3504R1-R27), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-a6545c7ce7f2b3d1892707ff1210f8a83ada7adf4892e1d49b8aad28490ec9f4R1-R23), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-a7d1fd059e6984fa35ba4ffe59c68ae224c798ad06e32d1467f2955d810480f4R1-R7), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR1028-R1037), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR25089-R25111))
*  Update the `viem` package version and its dependencies, which are used by the `@pancakeswap/chains` package and the `@safe-global/safe-gateway-typescript-sdk` package ([link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR2021-R2023), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4590-R4603), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4609-R4622), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR7353-R7357), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR7371-R7374), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL7734-R7756), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL7752-R7806), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL8005-R8043), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR10488-R10492), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR12338-R12351), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR18019-R18025))
*  Update the `@noble/hashes` package version, which is used by the `@solana/web3.js` and the `@solana/spl-token` packages ([link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4590-R4603), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4609-R4622), [link](https://github.com/pancakeswap/pancake-frontend/pull/7840/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL8005-R8043))

### CODE REVIEW

File: packages/chains/package.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  "prettier": {
+    "printWidth": 120,
+    "semi": false,
+    "singleQuote": true
+  },
+  "devDependencies": {
+    "tsup": "^6.7.0"
+  },
+  "dependencies": {
+    "viem": "^1.10.9"
```
Comment (chefjackson) [inline packages/chains/package.json:48]: let's bump all other viem versions as well if no peer dependency conflict

File: packages/chains/package.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist"
+  },
+  "engines": {
+    "node": ">=10"
+  },
+  "prettier": {
+    "printWidth": 120,
+    "semi": false,
+    "singleQuote": true
+  },
```
Comment (chefjackson) [inline packages/chains/package.json:43]: I think we have config under the root directory

File: packages/chains/src/chains.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  polygonZkEvmTestnet,
+  arbitrum,
+  arbitrumGoerli,
+  scrollSepolia,
+  scrollTestnet,
+  linea,
+  lineaTestnet,
+  base,
+  baseGoerli,
+} from 'viem/chains'
```
Comment (chefjackson) [inline packages/chains/src/chains.ts:20]: is there a reason for this export? If we need them somewhere else we can simply import from viem?

Comment (ChefJerry) [reply packages/chains/src/chains.ts:19]: in my view, this pkg is the single source of truth for pancakeswap chains, so all the chain define should in this pkg, include the predefine chain(directly reexport from viem), and global custom chain

File: packages/chains/package.json (hunk 1 above)
Comment (ChefJerry) [reply packages/chains/package.json:43]: done

File: packages/chains/src/chains.ts (hunk 3 above)
Comment (chefjackson) [reply packages/chains/src/chains.ts:19]: we could define our own typed chain object later on if that's the case, and we don't have to follow the naming convention and type definition from viem.

Let's discuss the things we need for the chain type after setting up the pkg and for now let's not re-export from viem

File: packages/swap-sdk/package.json [hunk 4]
```diff
@@ -40,7 +40,7 @@
     "tiny-invariant": "^1.1.0",
     "tiny-warning": "^1.0.3",
     "toformat": "^2.0.0",
-    "viem": "^1.2.9"
+    "viem": "^1.10.9"
```
Comment (chefjackson) [inline packages/swap-sdk/package.json:43]: let's add the `@pancakeswap/chains` as one of the dependencies to sdk. We can re-export the `ChainId` from our new pkg and mark it as deprecated

**Comment by chefjackson** [comment]
pls also help add changesets to other pkgs as they have dependency update

File: packages/chains/src/chains.ts (hunk 3 above)
Comment (ChefJerry) [reply packages/chains/src/chains.ts:19]: any examples of our chain object type? as far as I discovered, it used like the same as viem
https://github.com/pancakeswap/pancake-frontend/blob/c899962c860f6e9d4aaf90ff66822a21613911fa/apis/farms/src/provider.ts#L17-L96

Comment (chefjackson) [reply packages/chains/src/chains.ts:19]: we may extend on top of that or we can choose to have our own type. For example we may add subgraph links to the config

Comment (chefjackson) [reply packages/chains/src/chains.ts:19]: and probably we don't even need to export the viem related chain config as we may provide the viem client out of the box, e.g.

```typescript
import { viemProviders }  from '@pancakeswap/chains'
// or maybe
import { getViemClient }  from '@pancakeswap/chains`

const client = getViemClient({ chainId })
```

so far the most common usecase for the viem chain configs are creating public clients. We can wrap it up so we don't need to define the providers all over the place

Comment (chefjackson) [reply packages/chains/src/chains.ts:19]: and our config could be like something else
```typescript
const ethereum = {
  rpcUrls: [],
  subgraphs: {
    v3: '',
    v2: '',
    block: '',
  },
  ...
}
```

of course if we need to export our extended viem chain configs we can surely do, although I think no need for now
```typescript
import { mainnet as defaultMainnet } from 'viem/chains'

const mainnet = {
  ...defaultMainnet,
  // extended by our own config maybe
  rpcUrls: {
    // read from our own rpc url configs
  },
}

export const viemChains = {
  mainnet
}
```

Comment (chefjackson) [reply packages/chains/src/chains.ts:19]: we can leave this for further discussion for now. Let's push out the 0.0.1 version only including the `ChainId` first

File: packages/v3-sdk/package.json [hunk 5]
```diff
@@ -25,7 +25,6 @@
     "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist"
   },
   "dependencies": {
-    "viem": "^1.2.9",
     "@pancakeswap/sdk": "workspace:*",
```
Comment (chef-jojo) [inline packages/v3-sdk/package.json:28]: Wish list: remove v3 depending on v2 sdk

File: packages/swap-sdk/src/index.ts [hunk 6]
```diff
@@ -10,5 +10,6 @@ export * from './ether'
 export * from './fetcher'
 
 export * from '@pancakeswap/swap-sdk-core'
+export { ChainId } from '@pancakeswap/chains'
```
Comment (chefjackson) [inline packages/swap-sdk/src/index.ts:13]: Let's add a `@deprecated` mark here

File: packages/utils/package.json [hunk 7]
```diff
@@ -7,20 +7,20 @@
   "private": true,
   "peerDependencies": {
     "@pancakeswap/localization": "workspace:*",
-    "viem": "^1.2.9",
     "date-fns": "^2.29.3",
+    "lodash": "^4.17.21",
     "styled-components": "^6.0.7",
-    "lodash": "^4.17.21"
+    "viem": "^1.2.9"
```
Comment (chefjackson) [inline packages/utils/package.json:13]: inconsistent version here

Comment (ChefJerry) [reply packages/utils/package.json:13]: any `pnpm` cli method about up peerDeps version? cannot found on document

Comment (ChefJerry) [reply packages/utils/package.json]: just add again ...

**Review by chefjackson — APPROVED** [review]

**Pull request merged**